nemo/export/trt_llm/nemo_ckpt_loader/nemo_file.py

```diff
@@ -77,69 +79,68 @@ def __init__(self, path: Union[Path, TarPath]) -> None:
+    for key in scales:
+        if '.decoder.layers.0' not in key:
+            continue
 
+        # Key has a structure "model.decoder.layers.<layer_number>.<rest>"
+        decomposed = key.split('.')
+        layer_num_idx = 3
 
-def retrieve_scale(bytes: BytesIO) -> Optional[torch.Tensor]:
-    bytes.seek(0)
-    extra_state = torch.load(bytes)
-    if not extra_state or 'scale_fwd' not in extra_state:
-        return None
-    return extra_state['scale_fwd'].cpu()
+        # Merges scales from "model.decoder.layers.<layer_num>.<rest>" to
+        # larger dimensional tensor with "model.decoder.layers.<rest>" key
+        combined = []
+        layer_num = 0
+        decomposed[layer_num_idx] = str(layer_num)
+        while (scale := scales.get('.'.join(decomposed))) is not None:
+            combined.append(scale)
+            layer_num += 1
+            decomposed[layer_num_idx] = str(layer_num)
 
+        del decomposed[layer_num_idx]
+        combined_scales['.'.join(decomposed)] = torch.stack(combined)
 
-def load_scales_from_bytes(bytes_list: List[BytesIO]) -> Optional[torch.Tensor]:
-    scales = []
-    for bytes in bytes_list:
-        scale = retrieve_scale(bytes)
-        if scale is None:
-            return None
-        scales.append(scale)
-    return torch.stack(scales)
+    return state_dict | combined_scales
 
 
-def load_scaling_factors(state_dict: dict, basename: str, size: int) -> Optional[torch.Tensor]:
-    keynames = [f'{basename}/shard_{layer}_{size}' for layer in range(size)]
-    bytes_list = [state_dict[keyname][0] for keyname in keynames]
-    return load_scales_from_bytes(bytes_list)
+def rename_extra_states(state_dict: dict) -> dict:
+    mcore_extra_states = {}
 
+    for key, value in state_dict.items():
+        if EXTRA_STATE not in key:
+            continue
 
-def filter_experts_extra_states(state_dict: dict):
-    pattern = (
-        r'(model|module)\.decoder\.layers\.mlp\.experts\.experts\.linear_fc\d+\._extra_state/shard_\d+\.\d+_\d+\.\d+'
-    )
-    return {k: v for k, v in state_dict.items() if not re.fullmatch(pattern, k)}
+        # extra state key has a form: "<base_key>._extra_state/shard_<layer_num>_<num_layers>"
+        key_base, shard_key = key.split('/')
+        if '_' not in shard_key:
+            continue
 
+        shard_layer = shard_key.split('_')[1]
+        if not shard_layer.isnumeric():
+            continue
 
-def standarize_distributed_scaling_factors(state_dict: dict) -> dict:
-    while key := get_extra_state_key(state_dict):
-        basename, size = unpack_extra_state_key(key)
-        scaling_factors = load_scaling_factors(state_dict, basename, size)
-        if scaling_factors is not None:
-            state_dict[basename + '.scale_fwd'] = scaling_factors
-        state_dict = clear_loaded_extra_states(state_dict, basename)
+        mcore_key = key_base.replace("layers", f"layers.{shard_layer}")
+        if isinstance(value, list) and len(value) == 1:
```

Comment:
What are possible types of `value` if not only `List` (of different lengths)?

Reply:
The extra states `value` is:
* a `BytesIO` object, when loaded from `zarr`
* a one-element list with a `BytesIO` object, when loaded from `torch_dist`

I'll add a comment here